### PR TITLE
feat: js style import support

### PR DESCRIPTION
**Pull request by lifeart**
### Octane imports style

```hbs
import BasicDropdown from 'ember-basic-dropdown/components/basic-dropdown';
import { BasicDropdown as SameDropdown } from 'ember-basic-dropdown/components';
import PrivateComponent from 'src/ui/routes/items/-components/component-name';

--- hbs ---

<BasicDropdown />
<SameDropdown />
<PrivateComponent />
```

### Classic imports style

```hbs
{{import BasicDropdown from 'ember-basic-dropdown/components/basic-dropdown'}}
{{import SameDropdown from 'ember-basic-dropdown/components'}}
{{import PrivateComponent from 'src/ui/routes/items/-components/component-name'}}

<BasicDropdown />
<SameDropdown />
<PrivateComponent />
```

### CODE REVIEW

**Comment by NullVoxPopuli** [comment]
just tested this, and it works! :)

**Comment by knownasilya** [comment]
Any chance for a readme update?

**Comment by lifeart** [comment]
@knownasilya, yeah, nice catch! I'm planning to add more tests and then write down readme update

**Comment by davewasmer** [comment]
@lifeart thanks for all the work here! Just wanted to give you heads up that we've got a big deadline early next week, so it's likely we won't have a chance to review this until then. But don't take the silence as lack of interest!

**Comment by shanemcgraw** [comment]
Hey there @davewasmer - wondering if this PR is still in play? Would love to try this out in Octane :)

File: app/helpers/register-absolute-template-import.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/* global require, define */
+
+import { getOwner } from "@ember/application";
+import Helper from "@ember/component/helper";
+
+function generateHash(module, testName) {
+  var str = module + "\x1C" + testName;
+  var hash = 0;
+
+  for (var i = 0; i < str.length; i++) {
```
[the remark this answers is not in the document]
Comment (lifeart) [reply app/helpers/register-absolute-template-import.js]: https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript
This function generate string based hash.

File: lib/mu-utils.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    .replace(/[^a-zA-Z0-9-.@]+/g, " ")
+    .trim()
+    .split(" ")
+    .join("/");
+
+  // console.log('normalizeRightImportPart', right, importSt);
+  // do we need "/" ending?
+  //   if (!importSt.endsWith("/")) {
+  //     importSt = importSt + "/";
+  //   }
```
[the remark this answers is not in the document]
Comment (lifeart) [reply lib/mu-utils.js]: done

File: lib/mu-utils.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  // ember-basic-dropdown/components/basic-dropdown, let's handle addon's component's imports
+  // import { BasicDropdown } from 'ember-basic-dropdown/src/ui/components';
+  // import BasicDropdown from 'ember-basic-dropdown/src/ui/components/basic-dropdown';
+  //   if (!maybeImport.startsWith("/src/ui/") && maybeImport.includes("/src/ui/")) {
+  // const [addonName, componentName] = maybeImport.split("/src/ui/components/");
+  // return addonName + "/components/" + componentName;
+  //   }
+  return maybeImport;
```
[the remark this answers is not in the document]
Comment (lifeart) [reply lib/mu-utils.js]: done

File: app/helpers/register-absolute-template-import.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  return hex.slice(-8);
+}
+
+function getAppName(appInstance) {
+  if (appInstance.base && appInstance.base.name) {
+    return appInstance.base.name;
+  }
+  // TODO: would this work in 2.4+?
```
[the remark this answers is not in the document]
Comment (knownasilya) [reply app/helpers/register-absolute-template-import.js]: @lifeart just waiting on this to get it across the finish line. Thanks!

[the remark this answers is not in the document]
Comment (lifeart) [reply app/helpers/register-absolute-template-import.js]: @knownasilya ya, it's just copy-paste from https://github.com/lifeart/ember-ast-hot-load/blob/ea36a718e82e56d51dd3b38e4a640d792300cfbd/addon/instance-initializers/hot-loader-livereload-plugin.js#L87
I think we can go with it

**Comment by knownasilya** [comment]
@lifeart I thing the last thing holding this up is our hesitance with the MU part, since that's an ended experiment.

File: index.js [hunk 5]
```diff
@@ -1,7 +1,18 @@
-'use strict';
+"use strict";
 
 /* eslint-env node */
 
+const path = require("path");
```
Comment (knownasilya) [inline index.js]:
```suggestion
```

**Comment by lifeart** [comment]
Using https://github.com/lifeart/ember-language-server/issues/12 it's possible to implement ELS support for this feature, for example:

find text content in template

https://astexplorer.net/#/gist/97bd072e0190bc366b6e78d0801df678/84c54a1f052c2924c29a828773a9657d1dfb8374

parse it using babel parser

https://astexplorer.net/#/gist/4d939c5fe32758eaf10155e2c9140353/cdf596d302dd1e0aba625a41588dccdd7f14181b

resolve component name and return it back

**Pull request merged**